pkg/controller/oam/applicationconfiguration/component.go

```diff
@@ -0,0 +1,194 @@
+		return nil
+	}
+	var reqs []reconcile.Request
+	if match, namespaceName := isMatch(&appConfigs, object.GetName()); match {
+		reqs = append(reqs, reconcile.Request{NamespacedName: namespaceName})
+	}
+	return reqs
+}
+
+//IsRevisionDiff check whether there's any different between two component revision
+func (c *ComponentHandler) IsRevisionDiff(mt metav1.Object, curComp *v1alpha2.Component) (bool, int64) {
+	if curComp.Status.LatestRevision == nil {
+		return true, 0
+	}
+	oldRev, err := c.appsClient.ControllerRevisions(mt.GetNamespace()).Get(context.Background(), curComp.Status.LatestRevision.Name, metav1.GetOptions{})
+	if err != nil {
+		c.l.Info(fmt.Sprintf("get old controllerRevision %s error %v, will create new revision", curComp.Status.LatestRevision.Name, err), "componentName", mt.GetName())
+		return true, curComp.Status.LatestRevision.Revision
+	}
+	oldComp, err := UnpackRevisionData(oldRev)
+	if err != nil {
+		c.l.Info(fmt.Sprintf("Unmarshal old controllerRevision %s error %v, will create new revision", curComp.Status.LatestRevision.Name, err), "componentName", mt.GetName())
+		return true, oldRev.Revision
+	}
+
+	if reflect.DeepEqual(curComp.Spec, oldComp.Spec) {
+		return false, oldRev.Revision
+	}
+	return true, oldRev.Revision
+}
+
+// UnpackRevisionData will unpack revision.Data to Component
+func UnpackRevisionData(rev *appsv1.ControllerRevision) (*v1alpha2.Component, error) {
+	var err error
+	if rev.Data.Object != nil {
+		comp, ok := rev.Data.Object.(*v1alpha2.Component)
+		if !ok {
+			return nil, fmt.Errorf("invalid type of revision %s, type should not be %v", rev.Name, reflect.TypeOf(rev.Data.Object))
+		}
+		return comp, nil
+	}
+	var comp v1alpha2.Component
+	err = json.Unmarshal(rev.Data.Raw, &comp)
+	return &comp, err
+}
+
+func newTrue() *bool {
+	b := true
+	return &b
+}
+
+func (c *ComponentHandler) createControllerRevision(mt metav1.Object, obj runtime.Object) bool {
+	curComp := obj.(*v1alpha2.Component)
+	diff, curRevision := c.IsRevisionDiff(mt, curComp)
+	if !diff {
+		// No difference, no need to create new revision.
+		return false
+	}
+	nextRevision := curRevision + 1
+	revisionName := ConstructRevisionName(mt.GetName())
+	// set annotation to component
+	revision := appsv1.ControllerRevision{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: revisionName,
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					APIVersion: v1alpha2.SchemeGroupVersion.String(),
+					Kind:       v1alpha2.ComponentKind,
+					Name:       curComp.Name,
+					UID:        curComp.UID,
+					Controller: newTrue(),
+				},
+			},
+		},
+		Revision: nextRevision,
+		Data:     runtime.RawExtension{Object: curComp},
+	}
+	_, err := c.appsClient.ControllerRevisions(mt.GetNamespace()).Create(context.Background(), &revision, metav1.CreateOptions{})
+	if err != nil {
+		c.l.Info(fmt.Sprintf("error create controllerRevision %v", err), "componentName", mt.GetName())
+		return false
+	}
+	curComp.Status.LatestRevision = &v1alpha2.Revision{
```

Comment:
say the controller pod is killed here, we will have a dangling controller revision. Does it matter?

Reply:
It doesn't matter, finally the controllerrevision will be clean if the component is deleted